src/server/replica.cc

```diff
@@ -137,55 +137,58 @@ Replica::~Replica() {
 
 static const char kConnErr[] = "could not connect to master: ";
 
-bool Replica::Start(ConnectionContext* cntx) {
+error_code Replica::Start(ConnectionContext* cntx) {
   VLOG(1) << "Starting replication";
   ProactorBase* mythread = ProactorBase::me();
   CHECK(mythread);
 
+  RETURN_ON_ERR(cntx_.SwitchErrorHandler(absl::bind_front(&Replica::DefaultErrorHandler, this)));
+
+  auto check_connection_error = [this, &cntx](const error_code& ec, const char* msg) -> error_code {
+    if (cntx_.IsCancelled()) {
+      (*cntx)->SendError("replication cancelled");
+      return std::make_error_code(errc::operation_canceled);
+    }
+    if (ec) {
+      (*cntx)->SendError(absl::StrCat(msg, ec.message()));
+      cntx_.Cancel();
+      return ec;
+    }
+    return {};
+  };
+
   // 1. Resolve dns.
   VLOG(1) << "Resolving master DNS";
   error_code ec = ResolveMasterDns();
-  if (ec) {
-    (*cntx)->SendError(StrCat("could not resolve master dns", ec.message()));
-    return false;
-  }
+  RETURN_ON_ERR(check_connection_error(ec, "could not resolve master dns"));
+
   // 2. Connect socket.
   VLOG(1) << "Connecting to master";
   ec = ConnectAndAuth(absl::GetFlag(FLAGS_master_connect_timeout_ms) * 1ms);
-  if (ec) {
-    (*cntx)->SendError(StrCat(kConnErr, ec.message()));
-    return false;
-  }
+  RETURN_ON_ERR(check_connection_error(ec, kConnErr));
 
   // 3. Greet.
   VLOG(1) << "Greeting";
   state_mask_ = R_ENABLED | R_TCP_CONNECTED;
   last_io_time_ = mythread->GetMonotonicTimeNs();
   ec = Greet();
-  if (ec) {
-    (*cntx)->SendError(StrCat("could not greet master ", ec.message()));
-    return false;
-  }
+  RETURN_ON_ERR(check_connection_error(ec, "could not greet master "));
 
   // 4. Init basic context.
-  cntx_.Reset(absl::bind_front(&Replica::DefaultErrorHandler, this));
+  RETURN_ON_ERR(cntx_.SwitchErrorHandler(absl::bind_front(&Replica::DefaultErrorHandler, this)));
 
   // 5. Spawn main coordination fiber.
   sync_fb_ = MakeFiber(&Replica::MainReplicationFb, this);
 
   (*cntx)->SendOk();
-  return true;
-}
+  return {};
+}  // namespace dfly
 
 void Replica::Stop() {
   VLOG(1) << "Stopping replication";
   // Mark disabled, prevent from retrying.
-  if (sock_) {
-    sock_->proactor()->Await([this] {
-      state_mask_ = 0;  // Specifically ~R_ENABLED.
-      cntx_.Cancel();   // Context is fully resposible for cleanup.
-    });
-  }
+  state_mask_ = 0;  // Specifically ~R_ENABLED.
+  cntx_.Cancel();   // Context is fully resposible for cleanup.
```

Comment:
Either: 
- Lets store the replicas `ProactorBase` upon creation so we don't need to read sock
- Or lets add a mutex and switch the `sock_` ptr only on under it.

Reply:
Can't reset the `state_mask_` inside `CloseSocket` because we want to keep the `while(enabled)` loop even when a connection is reset. But I made it atomic and added a mutex. Hope I got it more right this time :)

Reply:
True, but we could set some kind of flag whether to reset it or not. Either way, the second option is fine as well